fuzzer/HttpEcho.cpp

```diff
@@ -0,0 +1,149 @@
+        }
+    };
+
+public:
+    HttpRequestTests()
+        : _pollServerThread("HttpServerPoll")
+        , _poller("HttpSynReqPoll")
+    {
+        _poller.runOnClientThread();
+
+        std::map<std::string, std::string> logProperties;
+        const auto log_level = std::getenv("LOG_LEVEL");
+        if (log_level)
+        {
+            Log::initialize("fuz", log_level ? log_level : "error", isatty(fileno(stderr)), false,
+                            logProperties);
+        }
+
+        std::shared_ptr<SocketFactory> factory = std::make_shared<ServerSocketFactory>();
+        int port = 9990;
+        for (int i = 0; i < 40; ++i, ++port)
+        {
+            // Try listening on this port.
+            _socket = ServerSocket::create(ServerSocket::Type::Local, port, Socket::Type::IPv4,
+                                           _pollServerThread, factory);
+            if (_socket)
+                break;
+        }
+
+        _localUri = "http://127.0.0.1:" + std::to_string(port);
+        _pollServerThread.startThread();
+        _pollServerThread.insertNewSocket(_socket);
+
+        _httpSession = http::Session::create(localUri());
+        if (!_httpSession)
+            throw std::runtime_error("Failed to create http::Session to " + localUri());
+
+        _httpSession->setTimeout(std::chrono::milliseconds(500));
+        _httpSession->setFinishedHandler(
+            [&](const std::shared_ptr<http::Session>&)
+            {
+                _completed = true;
+                return true;
+            });
+    }
+
+    ~HttpRequestTests()
+    {
+        _pollServerThread.stop();
+        _socket.reset();
+    }
+
+    const std::string& localUri() const { return _localUri; }
+    std::shared_ptr<http::Session> session() const { return _httpSession; }
+    SocketPoll& poller() { return _poller; };
+    bool isCompleted() const { return _completed; };
+    void resetCompleted() { _completed = false; };
+};
+
+#define CHECK(X)                                                                                   \
+    do                                                                                             \
+    {                                                                                              \
+        if (!(X))                                                                                  \
+        {                                                                                          \
+            assert(!(X));                                                                          \
+            __builtin_trap();                                                                      \
+        }                                                                                          \
+    } while (0)
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size)
+{
+    static HttpRequestTests test;
```

Comment:
It would be good to have an initial "seed" for this fuzzer, see e.g. fuzzer/httpresponse-data/load as an example to follow. Once you have that, you can update fuzzer/README to document how to to invoke the fuzzer.

Last: running such a fuzzer has the cost of 100% using one CPU core, so it would be useful if that cost brings benefits. Do you know if this fuzzer already found some problem in practice or is this more just a speculative one? Asking to find out if our Jenkins should have a 4th fuzzer running 7/24 or that would be a waste of resources.

Reply:
Very good point. I've ran the fuzzer multiple times and generated several billion runs. I found a couple of issues (one or two crashes and a deadlock) and fixed them.

How likely is it to find other issues? Not very likely, but if we make a change to the parser then it'd be good to re-run the fuzzer. Should we have a core dedicated 24/7? Probably not.

Reply:
Ah, in case it found a single real problem in the past, then I'm happy to have this in. :-)